flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/UpsertToRetractionProcessFunction.scala

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.table.runtime
+
+import org.apache.flink.api.common.state.{ValueState, ValueStateDescriptor}
+import org.apache.flink.api.java.typeutils.RowTypeInfo
+import org.apache.flink.configuration.Configuration
+import org.apache.flink.streaming.api.functions.ProcessFunction
+import org.apache.flink.table.api.StreamQueryConfig
+import org.apache.flink.table.runtime.aggregate.ProcessFunctionWithCleanupState
+import org.apache.flink.table.runtime.types.CRow
+import org.apache.flink.table.util.Logging
+import org.apache.flink.types.Row
+import org.apache.flink.util.Collector
+
+/**
+  * Function used to convert upsert to retractions.
+  *
+  * @param rowTypeInfo the output row type info.
+  * @param queryConfig the configuration for the query.
+  */
+class UpsertToRetractionProcessFunction(
```

Comment:
Ok, but then we can provide `LAST_NULLABLE_VALUE()` function and still re-use aggregation operator code. Even if we keep `LAST_NULLABLE_VALUE()` as an internal function that we do not expose anywhere. I would vote for exposing it (maybe in a follow up PR).

My main concern is again code duplication with aggregation operator. 

Now that I'm thinking about it, did we even need to implement our own UpsertToRetraction rules and RelNodes? Couldn't we re-use `Aggregate` `RelNode` with `LAST_NULLABLE_VALUE()` aggregation functions?

Reply:
I think your concern makes sense. However, I'm not sure if we can use the `LAST_NULLABLE_VALUE` function. According to the description of handling null, it is said: `aggregate functions ignore null values`. This means we can even skip processing the data. Some optimization can make use of it to gain better performance. Anything like this would be broken silently later. 
Again, I agree with you that it is good to reuse the aggregation code, so if you do think it is valuable to do. I can change the current implementation.

Reply:
> it is said: `aggregate functions ignore null values`

Can you point me where did you get this? Is this part of SQL standard? `COUNT` function definitely doesn't ignore null values, so I would assume we should be able to define custom aggregate functions that work in a similar fashion. [This seems to support this](
https://modern-sql.com/concept/null#aggregates):
> Most aggregate functions that return structured data (array_agg, json_objectagg, xmlagg) do not remove null values.
> ...
> For array_agg, it is even explicitly mentioned in SQL:2016-2: §10.9, General Rule 12gii, note 510.

Reply:
I haven't found related descriptions in SQL standard. I found these descriptions in some popular sql systems, such as[[1]](https://docs.microsoft.com/en-us/sql/t-sql/functions/aggregate-functions-transact-sql?view=sql-server-2017)[[2]](https://docs.oracle.com/database/121/SQLRF/functions003.htm#SQLRF20035).

>Except for COUNT, aggregate functions ignore null values. 

I'm fine if we decide to reuse aggregate operator and I prefer reuse all aggregate operator including planing instead only runtime part to make it more clear. What do you think?

Reply:
The only reference in SQL standard that I have found was this `SQL:2016-2: §10.9, General Rule 12gii, note 510.`, which explicitly states that `array_agg` can not ignore nulls. Also I haven't found in the standard  references to something stating the opposite, so I would be inclined to re-use aggregation node in planning as well if you are fine with that.

Next question would be wether to simply create an aggregation node or to extend either `Aggregate` or `FlinkLogicalAggregate` classes (to re-use all of their logic), but to keep the distinct name of the node `UpsertToRetraction`?

Reply:
OK. Maybe we can simply create a `FlinkLogicalAggregate` in `UpsertToRetractionConverter`, thus we can reuse the rules of Aggregate such as `FilterAggregateTransposeRule`?

Reply:
Just FYI. `FilterAggregateTransposeRule` should work with any class that extends `Aggregate` node, so if our `FlinkLogicalUpsertToRetraction` class would extend from `FlinkLogicalAggregate`, transpose rule should be fine.

Creating `FlinkLogicalAggregate` in `UpsertToRetractionConverter` is one option, but with a slight drawback of less easier to understand query plan. If logical/physical query plan has an `UpsertToRetraction` node it's obvious what it does. If there is an `Aggregate` node with some `LAST_NULLABLE_VALUE` function, it might be confusing "where did it come from?". However this point may be addressed by overriding `RelNode#getRelTypeName` in `FlinkLogicalAggregate`, so that the aggregation created by `UpsertToRetractionConverter` would return `"FlinkLogicalUpsertToRetraction"` name.

Reply:
Sorry for my typo. I meant to create `LogicalAggregate` in `UpsertToRetractionConverter`.  Some rules such as `AggregateProjectPullUpConstantsRule` only matches `LogicalAggregate`. Thus we can reuse more Aggregate rules for `UpsertToRetraction` node. 

To make our plan more understandable, we may can't override the `getRelTypeName` method. It is final in `AbstractRelNode`. However, we may override `explainTerms` and `toString` method to make the RelNode more understandable.
What do you think?

Reply:
Ops, I missed this `final` :/ Will overriding `explainTerms` or `toString` help with anything? I think calcite's explain (`org.apache.calcite.rel.externalize.RelWriterImpl#explain_`) hard codes using `getRelTypeName()` and we are using this for printing the plans, or am I missing something? 

Ok, please do as you decide is best. It would be nice to keep explained plan easy to understand, but if we just use plain aggregation for the sake of code simplicity, it's not the end of the world.

Reply:
OK, I will try to reimplement the PR and reuse the Aggregate node. I will let you know once it is ok. Thanks a lot for your valuable suggestions. :-)